src/Data/Machine/Mealy.purs

```diff
@@ -47,62 +49,115 @@ import Data.Tuple (Tuple(..), fst, snd, swap)
 import Data.Unfoldable (class Unfoldable, unfoldr)
 import Effect.Class (class MonadEffect, liftEffect)
 
+-- | Mealy is a finite state machine, where:
+-- |
+-- | - `f` is the effect under which we evaluate,
+-- | - `s` is the input state, and
+-- | - `a` is the output type.
 newtype MealyT f s a = MealyT (s -> f (Step f s a))
 
 runMealyT :: forall f s a. MealyT f s a -> s -> f (Step f s a)
 runMealyT (MealyT f) = f
 
--- | Transforms a Mealy machine running in the context of `f` into one running 
+-- | Transforms a Mealy machine running in the context of `f` into one running
 -- | in `g`, given a natural transformation from `f` to `g`.
 hoistMealyT :: forall f g s . Functor g => (f ~> g) -> MealyT f s ~> MealyT g s
 hoistMealyT f2g (MealyT goF) = MealyT goG
   where goG s = hoistStep f2g <$> f2g (goF s)
 
+-- | Step is the core for running machines. Machines can either stop
+-- | via the `Halt` constructor, or emit a value and recursively
+-- | construct the rest of the machine.
 data Step f s a = Emit a (MealyT f s a) | Halt
 
+-- | Transforms a step running in the context of `f` into one running
+-- | in `g`, given a natural transformation from `f` to `g`.
 hoistStep :: forall f g s . Functor g => (f ~> g) -> Step f s ~> Step g s
 hoistStep f2g (Emit v nxt) = Emit v (hoistMealyT f2g nxt)
 hoistStep _   Halt         = Halt
 
-
+-- | Sources are machines with trivial `Unit` input value.
 type Source f a = MealyT f Unit a
+
+-- | Sinks are machines with trivial `Unit` output values.
 type Sink f s = MealyT f s Unit
 
+-- | Wrap an effectful value into a source. The effect will be repeated
+-- | indefinitely.
+-- |
+-- | For example, generating ten instances of the value 1:
+-- | ```purescript
+-- | take 10 $ source (pure 1)
+-- | ```
 source :: forall f s. (Monad f) => f s -> Source f s
 source src =  mealy $ \_ -> flip Emit (source src) <$> src
 
+-- | Sinks are 'terminator nodes' in machines. They allow for an
```

Comment:
I think this part of the module comment might be better-suited for the `Sink` type (describing its purpose), and the comment for this function could focus on construction ("Construct a machine which executes an effectful computation on its inputs")